### PR TITLE
Pipes - Iuliia - Muncher

### PR DESCRIPTION
**Pull request by julalam**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| How did you go about exploring the Edamam API, how did you try querying the API? | I used postman to test different queries after exploring edamam website. |
| Describe your API Wrapper.  How did you decide on the methods you created? | In my API Wrapper I used two main methods: search and show recipe in order to fulfill the requirements. The search method is searching Edamam API with given query and using private  method creates new instances of class Recipe. Show recipe method doing similar actions for one specific recipe. |
| Describe an edge case or failure case test you wrote for your API Wrapper. | I tested for raising an error if passed invalid app_id and app_key in search recipes method |
| Explain how VCR aids in testing an API. | I used VCR cassettes to record the response from API and use this record in future tests |
| What is the Heroku URL of your deployed application? | https://ic-muncher.herokuapp.com/ |

### CODE REVIEW

File: test/controllers/recipes_controller_test.rb [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+      VCR.use_cassette("recipes") do
+        search = {
+          query: ""
+        }
+
+        get recipes_path, params: search[:query]
+
+        must_respond_with :success
+      end
+    end
```
Comment (droberts-sea) [inline test/controllers/recipes_controller_test.rb:36]: There are some interesting test cases around paging that you're missing here:

1. Does the page still render if paging parameters are passed?
1. What if the paging parameters are invalid (negative number, etc)?
1. What if the paging parameters take you past the end of the available search results?

File: test/lib/edamam_api_wrapper_test.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+      VCR.use_cassette("no recipes") do
+        proc { EdamamApiWrapper.search_recipes("fish", "invalid_app_id", "invalid_app_key") }.must_raise EdamamApiWrapper::ApiError
+      end
+    end
+  end
+
+  describe "show_recipe" do
+    it "returns data for a recipe" do
+      VCR.use_cassette("recipe") do
+        uri = "637913ec61d9da69eb451818c3293df2"
```
Comment (droberts-sea) [inline test/lib/edamam_api_wrapper_test.rb:26]: I would also like to see a negative test case here - what happens if the URI is invalid?

File: test/lib/edamam_api_wrapper_test.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        recipes.length.must_be :>, 0
+        recipes.each do |recipe|
+          recipe.must_be_kind_of Recipe
+        end
+      end
+    end
+
+    it "raises an ApiError when the credentials are bad" do
+      VCR.use_cassette("no recipes") do
+        proc { EdamamApiWrapper.search_recipes("fish", "invalid_app_id", "invalid_app_key") }.must_raise EdamamApiWrapper::ApiError
```
Comment (droberts-sea) [inline test/lib/edamam_api_wrapper_test.rb:18]: What about for a search with no hits?

**Comment by droberts-sea** [comment]
# API Muncher
## What We're Looking For

| Feature | Feedback    |
| :------------- | :------------- |
| **Core Requirements** |   |
| Git hygiene | yes |
| Comprehension questions	| yes |
| **General** |  |
| Rails fundamentals (RESTful routing, use of named paths) | yes |
| Semantic HTML | yes |
| Errors are reported to the user | yes |
| API Wrapper to handle the API requests | yes |
| Controller testing | yes - missing tests around paging |
| Lib testing | yes - could use some expansion |
| Search Functionality | yes |
| List Functionality | yes |
| Show individual item functionality (link to original recipe opens in new tab) | yes |
| **Styling** |  |
| Foundation Styling for responsive layout | yes |
| List View shows 10 items at a time/pagination | yes |
| The app is styled to create an attractive user interface | yes |
| **API Features** | |
| The App attributes Edaman | yes |
| The VCR casettes do not contain the API key | yes |
| **External Resources** |  |
| Link to deployed app on Heroku | yes |
|  **Overall** | Excellent job overall! This site is attractive and responsive, easy to navigate and for the most part well-tested. Keep up the hard work! |
